Carbon tutorial web components (#4347)
carbon-design-system/carbon-website: src/pages/developing/web-components-tutorial/step-2.mdx

---
title: 2. Building pages
description:
  Welcome to Carbon! This tutorial will guide you in creating a Vanilla JS/HTML
  app using Web Components from the Carbon Design System.
tabs: ['Overview', 'Step 1', 'Step 2', 'Step 3', 'Step 4', 'Step 5']
---

import Preview from 'components/Preview';

<PageDescription>

Now that we have our app using the UI Shell, it's time to build a few static
pages. In this step, we'll become comfortable with the Carbon grid and various
Carbon components.

</PageDescription>

<AnchorLinks>

<AnchorLink>Fork, clone and branch</AnchorLink>
<AnchorLink>Add landing page grid</AnchorLink>
<AnchorLink>Build landing page</AnchorLink>
<AnchorLink>Style landing page</AnchorLink>
<AnchorLink>Add repo page grid</AnchorLink>
<AnchorLink>Build repo page</AnchorLink>

> <AnchorLink>Push to Github</AnchorLink>

Reviewer: ```suggestion
<AnchorLink>Push to GitHub</AnchorLink>
```

information review
[2x Grid](https://carbondesignsystem.com/elements/2x-grid/overview/) on the
Carbon Design System website. If apply classes manually it is also well worth
checking out the
[CSS Grid Demo](https://carbon-elements.netlify.app/grid/examples/css-grid/) to

</InlineNotification>

Then import the grid styles in `style.scss`.

```scss path=style.scss
@use '@carbon/styles/scss/grid';
```

In order to use the grid, we need to wrap everything in a `cds--css-grid`.
Because we're building with the new CSS Grid, we won't be using typical rows. We
have used a combination of `cds--css-grid-column` classes to create our layout.

The CSS Grid is a 16 column grid. The column widths and positions are specified
using a combination of `cds--BP:col-span-N` and `cds--BP:col-start-N`classes.
Where BP is the Carbon breakpoint and N is column count or starting position.
For example, `<div class="cds-lg:col-start-9 cds-lg:col-span-8" >` means the
column will start at column 9 and span 8 columns when the large (lg) breakpoint
is active. large breakpoint.

https://carbon-elements.netlify.app/grid/examples/css-grid/

We've included the designs for this tutorial app in the `design.figma` file

> found as a top-level file in the `carbon-tutorial` repository. But, if you don't

Reviewer 2: I couldn't find the figma file..

Author: Fixed

> have Sketch installed and available to inspect the design, we'll provide

Reviewer: ```suggestion
have Figma installed and available to inspect the design, we'll provide
```

screenshots.

![Landing page grid](../shared/step-2/images/landing-layout.png)

<Caption>Landing page grid</Caption>

<InlineNotification>

**Pro tip:** `SHIFT-G` toggles the layout in Figma.

</InlineNotification>

## Build landing page

We'll start adding HTML elements and components by row.

### First row

<Row>
<Column colLg={8}>

![Banner vertical spacing](../shared/step-2/images/landing-r1-spacing.png)

<Caption>Banner vertical spacing</Caption>

</Column>
</Row>

In our first row we'll need a `Breadcrumb` component. First, let's import the
components we need.

```javascript path=main.js
import '@carbon/web-components/es/components/breadcrumb/index';
```

We can now add our component to the first row, replace the content of

> `page-landing__banner` with:

Reviewer 2: ```suggestion
the `div` with class `page-landing__banner` with:
```

.page-landing__illo {
  max-width: 100%;
  float: inline-end;
  height: auto;
}

@include breakpoint-down(md) {
  .page-landing__illo {
    max-width: 528px;
    width: 100%;
    height: 100%;
    float: inline-start;
  }
}

.page-landing__tabs {
  margin: (-1 * $spacing-08) 0 $spacing-08;
}

.page-landing__tab-content {
  padding-block: $spacing-06;
}

.page-landing__subheading {
  @include type-style('productive-heading-03');

  font-weight: 600;
}

.page-landing__p {
  @include type-style('productive-heading-03');
  margin-top: $spacing-06;
  margin-bottom: $spacing-08;
}
```

### Third row

```scss path=style.scss
.page-landing__r3 {
  padding-block: $spacing-09;
  background: $layer-01;
}
```

Ta-da! You should see a step 2 complete landing page! Now we can move on to the
repo page.

## Build repo page

### Add a grid to contain our content

Now in our `repositories` page we will first add a grid wrapping
`REPOSITORIES PAGE`

```html path=repositories.html
<div class="page page-repositories cds--css-grid cds--css-grid--full-width">
  <div class="repo-page__r1 cds--css-grid-column cds--col-span-100">
    REPOSITORIES PAGE
  </div>
</div>

> :

Reviewer 2: ```suggestion
```

```

Add a minimal amount of styling to move our content away from the edge of the
page in `style.scss`.

```scss style.scss
.repo-page__r1 {
  padding-block: $spacing-05;
}
```

### Adding a table

Before we can add the table we need to import the web component. As this is only

> used in our `repositories.htnl` page lets create a new script file `repos.js`

Reviewer 2: ```suggestion
used in our `repositories.html` page lets create a new script file `repos.js`
```

and then add.

```javascript path=repos.js
import '@carbon/web-components/es/components/data-table/index.js';
```

We need to include this file in `repositories.html` which we can do by adding
the following next to the script tag that includes `main.js`.

```html path=repositories.html
<script type="module" src="/repos.js"></script>
```

> Next we add the table header and column titles.

Reviewer 2: ```suggestion
Next we add the table header and column titles to replace the text `REPOSITORIES PAGE`
```

  <cds-table-cell>Links</cds-table-cell>
</cds-table-row>
<cds-table-expanded-row>Repo description</cds-table-expanded-row>
<cds-table-row>
  <cds-table-cell>Repo 2</cds-table-cell>
  <cds-table-cell>Date</cds-table-cell>
  <cds-table-cell>Date</cds-table-cell>
  <cds-table-cell>123</cds-table-cell>
  <cds-table-cell>456</cds-table-cell>
  <cds-table-cell>Links</cds-table-cell>
</cds-table-row>
<cds-table-expanded-row>Repo description</cds-table-expanded-row>
<cds-table-row>
  <cds-table-cell>Repo 3</cds-table-cell>
  <cds-table-cell>Date</cds-table-cell>
  <cds-table-cell>Date</cds-table-cell>
  <cds-table-cell>123</cds-table-cell>
  <cds-table-cell>456</cds-table-cell>
  <cds-table-cell>Links</cds-table-cell>
</cds-table-row>
<cds-table-expanded-row>Repo description</cds-table-expanded-row>
```

### Using HTML templates

With the app running we can see that the repositories page now hosts a table.
However, it is not realistic to populate a table with hard coded data way so
we'll refactor to build the table from data.

This involves the use of HTML Templates, take a look at
[W3 Schools](https://www.w3schools.com/tags/tag_template.asp) if you need a
quick refresh.

> In `index.html` remove the contents of the `<cds-table>` tag and return it to

Reviewer 2: ```suggestion
In `index.html` remove the contents of the `<cds-table-body>` tag and return it to
```

`Table body goes here`.

> Then before the end of the HTML document add the following to define our table

Reviewer 2: ```suggestion
Then before the end of the html tag and after the body closing tag add the following to define our table
```

    updated: 'Date',
    openIssues: 123,
    stars: 456,
    links: 'Links',
    expansion: 'Row description',
  },
  {
    name: 'Repo C',
    created: 'Date',
    updated: 'Date',
    openIssues: 123,
    stars: 456,
    links: 'Links',
    expansion: 'Row description',
  },
];
```

Next we create the function `updateTable` and make a call to it to populate add
our rows.

```javascript path=repos.js
const updateTable = () => {
  const tableRowTemplate = document.querySelector(
    'template#template--table-row'
  );
  const tableBody = document.querySelector('cds-table-body');
  if (tableBody && tableRowTemplate) {
    tableBody.innerHTML = '';
    // iterate over data and render rows
    data.forEach((row) => {
      let newRow = tableRowTemplate.content.cloneNode(true);
      const keys = Object.keys(row);
      keys.forEach((key) => {
        const keyEl = newRow.querySelector(`[key="${key}"]`);
        keyEl.innerHTML = row[key];
      });
      tableBody.appendChild(newRow);
    });
  }
};

updateTable();
```

Verify that the table rows are being generated by changing the data and running
the app.

Congratulations! We've now created our static repo page!

> ## Push to Github

Reviewer: ```suggestion
## Push to GitHub
```
